Sources/X509/X509BaseTypes/TimeCalculations.swift

```diff
@@ -0,0 +1,263 @@
+    static var daysPer100Years: Int64 {
+        365 * 100 + 24
+    }
+
+    @inlinable
+    static var daysPer4Years: Int64 {
+        365 * 4 + 1
+    }
+
+    @inlinable
+    static func daysInMonth(_ month: Int) -> Int64 {
+        // This may seem weird, but these months are indexed from _March_.
+        // Thus, month 0 is March, month 11 is February.
+        switch month {
+        case 0, 2, 4, 5, 7, 9, 10:
+            return 31
+        case 1, 3, 6, 8:
+            return 30
+        case 11:
+            return 29
+        default:
+            fatalError()
+        }
+    }
+
+    @inlinable
+    var utcDateFromTimestamp: (year: Int, month: Int, day: Int, hours: Int, minutes: Int, seconds: Int) {
+        let secs = self - .leapoch
+        var (days, remsecs) = secs.quotientAndRemainder(dividingBy: 86400)
+
+        // Tolerate negative values
+        if remsecs < 0 {
+            // Unchecked is safe here: we know remsecs is negative, and we know
+            // days cannot be Int64.min
+            remsecs &+= 86400
+            days &-= 1
+        }
+
+        var (qcCycles, remdays) = days.quotientAndRemainder(dividingBy: .daysPer400Years)
+        if remdays < 0 {
+            // Same justification here for unchecked arithmetic
+            remdays &+= .daysPer400Years
+            qcCycles &-= 1
+        }
+
+        // Unchecked arithmetic here is safe: the subtraction is all known values (4 - 1),
+        // and the multiplication cannot exceed the original value of remdays.
+        var cCycles = remdays / .daysPer100Years
+        if cCycles == 4 { cCycles &-= 1 }
+        remdays &-= cCycles &* .daysPer100Years
+
+        // Unchecked arithmetic here is safe: the subtraction is all known values (25 - 1),
+        // and the multiplication cannot exceed the original value of remdays.
+        var qCycles = remdays / .daysPer4Years
+        if qCycles == 25 { qCycles &-= 1 }
+        remdays &-= qCycles &* .daysPer4Years
+
+        // Unchecked arithmetic here is safe: the subtraction is all known values (4 - 1),
+        // and the multiplication cannot exceed the original value of remdays.
+        var remyears = remdays / 365
+        if remyears == 4 { remyears &-= 1 }
+        remdays &-= remyears &* 365
+
+        // Unchecked multiplication here is safe: each of these has earlier been multiplied by
+        // a much larger number and we didn't need checked math then, so we don't need it now.
+        var years = remyears + (4 &* qCycles) + (100 &* cCycles) + (400 &* qcCycles)
+
+        var months = 0
+        while Int64.daysInMonth(months) <= remdays {
+            remdays -= Int64.daysInMonth(months)
+            months += 1
+        }
+
+        // Now we normalise the months count back to starting in January.
+        // Safe to do unchecked subtraction here becuase for all numbers 10 or
+        // larger we can subtract by 12.
+        if months >= 10 {
+            months &-= 12
+            years += 1
```

Comment:
This can be unchecked as well. The previous year calculations says that it multiplies with number which are much smaller than the previous ones so adding just one will be safe.
```suggestion
            years &+= 1
```

Reply:
I don't think that argument holds: while it was true for each element _individually_, it may not be true for the sum of them, which is why the addition between those values was not unchecked.